pre_commit/output.py

```diff
@@ -30,3 +42,16 @@ def write_line_b(
 
 def write_line(s: Optional[str] = None, **kwargs: Any) -> None:
     write_line_b(s.encode() if s is not None else s, **kwargs)
+
+
+@contextlib.contextmanager
+def paused_stdout() -> Generator[None, None, None]:
+    redirected_output = io.TextIOWrapper(io.BytesIO())
+    with contextlib.redirect_stdout(redirected_output):
+        yield
+        # We need to hold this lock through resetting stdout _and_ writing the saved contents,
+        # because otherwise another thread might write to stdout before we can write the buffered
+        # stdout, resulting in out-of-order output.
+        stdout_lock.acquire()
+    write_b(cast(io.BytesIO, redirected_output.buffer).getvalue(), sys.stdout.buffer)  # Supply buffer here so we don't deadlock.
+    stdout_lock.release()
```

Comment:
Ug, I kept looking for a way to use the lock with `with:` but it really doesn't seem possible haha.

Reply:
Yep. Spent a while thinking about that haha.

Reply:
Well, we actually could if we saved the old stdout, but that felt more gross. /shrug